data_dir/latte/vector_search/vector_search_recall.rn

```diff
@@ -0,0 +1,277 @@
+}
+
+////////////////////
+// User functions //
+////////////////////
+
+pub async fn insert(db, i) {
+    // Get the line for this iteration (dataset loaded in prepare phase)
+    let line = db.data.dataset_lines[i];
+
+    // Parse the line to extract id and vector
+    let parsed = parse_dataset_line_into_id_and_vector_string(line)?;
+    let vector_values = parse_bracketed_comma_separated_vector(parsed.vector_string)?;
+
+    let params = #{ "id": parsed.id, "vector": vector_values };
+    db.execute_prepared(P_STMT.INSERT.NAME, params).await?;
+}
+
+pub async fn validate_average_recall(db, i) {
+    // Get a random vector from test_data
+    let random_idx = latte::hash(i) % db.data.test_vectors.len();
+    let vector_string = db.data.test_vectors[random_idx].trim();
+    let ground_truth_ids = db.data.ground_truth_ids[random_idx];
+
+    // Parse vector string into list of floats
+    let vector = parse_bracketed_comma_separated_vector(vector_string)?;
+
+    // Execute ANN query with the random vector
+    let params = #{ "vector": vector };
+    let rows = db.execute_prepared_with_result(P_STMT.GET.NAME, params).await?;
+
+    if rows.is_empty() {
+        println!("WARNING: Query #{} returned no results!", i);
+        return Ok(());
+    }
+
+    // Extract actual IDs from query results
+    let actual_ids = [];
+    for row in rows {
+        actual_ids.push(row.id);
+    }
+
+    // Take first N IDs of ground truth (matching ANN_LIMIT)
+    let expected_ids = [];
+    for j in 0..ANN_LIMIT {
+        expected_ids.push(ground_truth_ids[j]);
+    }
+
+    // Count how many actual IDs are in expected IDs
+    let elements_present = 0;
+    for actual_id in actual_ids {
+        if contains(expected_ids, actual_id) {
+            elements_present = elements_present + 1;
+        }
+    }
+
+    // Calculate recall for this query
+    let query_recall = 0.0;
+    if !actual_ids.is_empty() {
+        query_recall = elements_present as f64 / actual_ids.len() as f64;
+    }
+
+    // Group recalls by chunks of 1000 queries for periodic validation
+    let chunk_key = (i / 1000).to_string();
+
+    // Get or create the recalls array for this chunk
+    if !db.data.recalls_map.contains_key(chunk_key) {
+        db.data.recalls_map[chunk_key] = [];
+    }
+
+    db.data.recalls_map[chunk_key].push(query_recall);
+
+    // When all requests per chunk executed, compute average recall for this chunk and assert it's above threshold
+    if (db.data.recalls_map[chunk_key].len()) % 1000 == 0 {
+        let recalls = db.data.recalls_map[chunk_key];
+
+        let sum = 0.0;
+        for r in recalls {
+            sum = sum + r;
+        }
+        let avg = sum / recalls.len() as f64;
+        println!("RECALL_PROGRESS: chunk={} avg_recall={:.3} threshold={}", chunk_key, avg, MIN_RECALL_THRESHOLD);
+        assert!(avg > MIN_RECALL_THRESHOLD);
+
+        // Remove the chunk from the map to free memory
+        db.data.recalls_map.remove(chunk_key);
```

Comment:
Do not delete key, assign empty list.
It will allow you to avoid race above when condition on the line 124 gets computed by parallel queries.